Added support of timestamp/date/time using curly brackets

matthewryanwells opened the pull request:
### Description
Added support for using timestamp/date/time using curly brackets in the following ways to fix failing date filters in PowerBI:
| Definition | Type |
| --- | --- |
| {ts 'YYYY-MM-DD hh:mm:ss[.SSS]'} | Timestamp |
| {timestamp 'YYYY-MM-DD hh:mm:ss[.SSS]'} | Timestamp |
| {d 'YYYY-MM-DD'} | Date |
| {date 'YYYY-MM-DD'} | Date |
| {t 'hh:mm:ss[.SSS]'} | Time |
| {time 'hh:mm:ss[.SSS]'} | Time |

 
### Issues Resolved
https://github.com/opensearch-project/sql/issues/364
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Yury-Fridlyand commented on core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:
```diff
@@ -1773,6 +1773,23 @@ public void testTimeFormatAndDateFormatReturnSameResult() {
   }
 
+  @Test
+  public void testTimestampBracket() {
+    FunctionExpression expr = DSL.time_format(
```
What do you test there?

Yury-Fridlyand commented on integ-test/src/test/java/org/opensearch/sql/legacy/QueryIT.java:
```diff
@@ -746,6 +746,7 @@ public void dateSearch() throws IOException {
     }
   }
 
+  @Ignore
```
Why?

acarbonetto replied:
This test will fail, since he's added syntax that makes this test run against the V2 engine.

Yury-Fridlyand replied:
As I understand test runs request in `json` format, so it should be executed by V1

matthewryanwells replied:
🔔 🔔 🔔 3 Points to Yury, the test still passes and is executed on V1, removed the `@ignore` in https://github.com/Bit-Quill/opensearch-project-sql/pull/297/commits/966efbb082b78fca450cd436641d3a3fe7a81842

Yury-Fridlyand commented on sql/src/main/java/org/opensearch/sql/sql/parser/AstExpressionBuilder.java:
```diff
@@ -9,6 +9,7 @@
 import static org.opensearch.sql.ast.dsl.AstDSL.between;
 import static org.opensearch.sql.ast.dsl.AstDSL.not;
 import static org.opensearch.sql.ast.dsl.AstDSL.qualifiedName;
+import static org.opensearch.sql.expression.function.BuiltinFunctionName.DATETIME;
```
typo or incomplete fix?

matthewryanwells replied:
typo, removed in https://github.com/Bit-Quill/opensearch-project-sql/pull/297/commits/966efbb082b78fca450cd436641d3a3fe7a81842

Yury-Fridlyand reviewed (commented):
So far so good. Please add
1. `{dt '...'}` and `{datetime ...}` too
2. UT for parser where you verify that AST from `select date '...'` and `select {d '...'}` are equal (and similar tests).
3. Doctests

acarbonetto commented on integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java:
```diff
@@ -1288,4 +1288,37 @@ protected JSONObject executeQuery(String query) throws IOException {
   }
+
+  @Test
+  public void testTimestampBracket() throws IOException {
```
add a test with microseconds too
```
ts '2020-09-16 17:30:00.123'
```

matthewryanwells replied:
added in https://github.com/Bit-Quill/opensearch-project-sql/pull/297/commits/966efbb082b78fca450cd436641d3a3fe7a81842

acarbonetto commented on integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java:
```diff
@@ -1288,4 +1288,37 @@ protected JSONObject executeQuery(String query) throws IOException {
+
+  @Test
+  public void testTimeBracket() throws IOException {
+    JSONObject result = executeQuery("select {time '17:30:00'}");
```
Add a test with microseconds too

matthewryanwells replied:
added in https://github.com/Bit-Quill/opensearch-project-sql/pull/297/commits/966efbb082b78fca450cd436641d3a3fe7a81842

GumpacG commented:
Could you add tests for invalid string literals? For example, `{ts "hello world"}` should throw an error.

Yury-Fridlyand commented:
Checkstyle fails
I think you need to update integ branch too

matthewryanwells commented:
> So far so good. Please add
> 
> 1. `{dt '...'}` and `{datetime ...}` too
> 2. UT for parser where you verify that AST from `select date '...'` and `select {d '...'}` are equal (and similar tests).
> 3. Doctests

1. We have decided to not add dt/datetime as PowerBI does not use that datatype
2. I have added unit tests and also added integration tests comparing the output of non bracketed to bracketed outputs
3. Add doctests

matthewryanwells commented:
> Could you add tests for invalid string literals? For example, `{ts "hello world"}` should throw an error.

Added tests for invalid inputs

matthewryanwells commented:
> Checkstyle fails I think you need to update integ branch too

Fixed checkstyle failures and integ branch has been updated

GumpacG commented on integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java:
```diff
@@ -1288,4 +1288,103 @@ protected JSONObject executeQuery(String query) throws IOException {
+      fails = true;
+    }
+
+    assertTrue(fails);
```
This can be improved to check for exception message. 
If we expect an exception to be thrown, `assertThrows` should be used.

matthewryanwells replied:
That's a great idea, I had no idea that existed. I have updated the integ branch and that change will appear on the upstream PR

GumpacG commented on integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java:
```diff
@@ -1288,4 +1288,103 @@ protected JSONObject executeQuery(String query) throws IOException {
+  }
+
+  private void queryFails(String query) {
+    Request request = new Request("POST", QUERY_API_ENDPOINT);
```
Why not use `executeQuery` like the rest of the tests?

matthewryanwells replied:
At the time I thought I had to do it this way because I didn't know of the existence of `assertThrows`, I have updated the code on the integ branch and removed this part as I had merged at the exact point you left this comment
